lib/koans/19_genservers.ex

```diff
@@ -0,0 +1,137 @@
+defmodule GenServers do
+  use Koans
+
+  @intro "GenServers"
+
+  defmodule BicycleLock do
+    use GenServer
+
+    #####
+    # External API
+
+    def start_server(init_password) do
+      # The __MODULE__ macro returns the current module name as an atom
+      GenServer.start_link(__MODULE__, init_password, name: __MODULE__)
+    end
+
+    def unlock(password) do
+      GenServer.call(__MODULE__, {:unlock, password})
+    end
+
+    def change_password(old_password, new_password) do
+      GenServer.cast(__MODULE__, {:change_password, old_password, new_password})
+    end
+
+    def owner_info do
+      GenServer.call(__MODULE__, :get_owner_info)
+    end
+
+    ####
+    # GenServer implementation
+    
+    def handle_call(:get_password, _from, current_password) do
+     {:reply, current_password, current_password} 
+    end
+
+    def handle_call(:get_bike_brand, _from, current_state) do
+      {:reply, "Tribe Bicycle Co.", current_state}
+    end
+
+    def handle_call(:get_bike_name, _from, current_state) do
+      {:reply, "CRMO Series", current_state}
+    end
+
+    def handle_call(:get_owner_info, _from, current_state) do
+      {:reply, {:ok, "Argh...Jack Sparrow's password is: #{current_state}"}, current_state}
+    end
+
+    def handle_call(:get_multi, _from, current_state) do
+      {:reply, {:ok, ["this", "is", "sparta"], 369, :hello_world}, current_state}
+    end
+
+    def handle_call(:name_check, _from, current_state) do
+      {:reply, "Congrats! Your process was successfully named.", current_state}
+    end
+    
+    def handle_call({:unlock, password}, _from, current_password) do
+      case password do
+        password when password === current_password ->
+          {:reply, {:ok, "Bicycle unlocked!"}, current_password}
+        _ -> 
+          {:reply, {:error, "Incorrect password!"}, current_password}
+      end
+    end
+    
+    def handle_cast({:change_password, old_password, new_password}, current_password) do
+      case old_password do
+        old_password when old_password == current_password ->
+          {:noreply, new_password}
+        _ ->
+          {:noreply, current_password} 
+      end
+    end
+  end
+
+  koan "Servers that are created and initialized successfully returns a tuple that holds the PID of the server" do
+    {:ok, pid} = GenServer.start_link(BicycleLock, nil)
+    assert is_pid(pid) == ___
+  end
+
+  koan "When starting a GenServer you can set it's initial state" do
+    {:ok, pid} = GenServer.start_link(BicycleLock, "Hey Arnold!")
+    assert GenServer.call(pid, :get_password) == ___
+  end
+
+  koan "The handle_call callback is synchronous so it will block until a reply is received" do
+    {:ok, pid} = GenServer.start_link(BicycleLock, "3kr3t!")
+    assert GenServer.call(pid, :get_password) == ___
```

Comment:
The fact that `call`s are synchronous are not really visible in this example. For us, it is clear that it must be synchronous if we want a return value, but a newcommer might miss this and end up confused.